src/Adapter/DoctrineItems.php

```diff
@@ -0,0 +1,390 @@
+    }
+
+    private function getCurrentVersion(ItemId $id, int $lockMode = LockMode::NONE) : ItemVersionNumber
+    {
+        $platform = $this->connection->getDatabasePlatform();
+
+        $sql = sprintf(
+            'SELECT item.current_version FROM %s WHERE item.id = :id %s',
+            $platform->appendLockHint("{$this->tableItems} item", $lockMode),
+            $this->getLockModeSql($lockMode)
+        );
+
+        $stmt = $this->connection->executeQuery($sql, ['id' => (string) $id]);
+        $current = $stmt->fetch(FetchMode::COLUMN);
+
+        if (false === $current) {
+            throw new ItemNotFound($id);
+        }
+
+        return ItemVersionNumber::create($current);
+    }
+
+    private function deleteItem(ItemId $id) : void
+    {
+        $this->connection->transactional(
+            function () use ($id) : void {
+                $this->connection->delete($this->tableItems, ['id' => (string) $id]);
+                $this->connection->delete($this->tableVersions, ['id' => (string) $id]);
+            }
+        );
+    }
+
+    private function deleteVersion(ItemId $id, ItemVersionNumber $current) : void
+    {
+        $this->connection->transactional(
+            function () use ($id, $current) : void {
+                $this->connection->update(
+                    $this->tableItems,
+                    ['current_version' => $current->previous()],
+                    ['id' => (string) $id]
+                );
+
+                $this->connection->delete(
+                    $this->tableVersions,
+                    ['id' => (string) $id, 'version' => $current->toInt()]
+                );
+            }
+        );
+    }
+
+    private function insertNewItem(ItemVersion $item) : void
+    {
+        if (1 !== $item->getVersion()->toInt()) {
+            throw new UnexpectedVersionNumber(
+                $item->getId(),
+                $item->getVersion(),
+                ItemVersionNumber::fromInt(1)
+            );
+        }
+
+        rewind($item->getContent());
+
+        $this->connection->transactional(
+            function () use ($item) : void {
+                $this->connection->insert(
+                    $this->tableVersions,
+                    [
+                        'id' => (string) $item->getId(),
+                        'version' => $item->getVersion()->toInt(),
+                        'content' => $item->getContent(),
+                        'hash' => $item->getHash(),
+                    ],
+                    ['content' => ParameterType::LARGE_OBJECT]
+                );
+
+                $this->connection->insert(
+                    $this->tableItems,
+                    [
+                        'id' => (string) $item->getId(),
+                        'current_version' => $item->getVersion()->toInt(),
+                    ]
+                );
+            }
+        );
+    }
+
+    private function insertVersion(ItemVersion $item) : void
```

Comment:
this is already called within a transaction (only call I could see), and the outer transaction is the one that really counts; so another "nested" transaction here doesn't seem to add safety

Reply:
Reverted in c73489e5f2ffaba3350ecc9079b6df594d20749d, @giorgiosironi still happy to merge?